tests/unit/client/auth/test_auth.py

```diff
@@ -0,0 +1,126 @@
+from tests.unit.util import execute_generator_requests
+
+
+def test_auth_refresh_on_expiration(
+    httpx_mock: HTTPXMock, test_token: str, test_token2: str
+) -> None:
+    """Auth refreshes the token on expiration."""
+    url = "https://host"
+    httpx_mock.add_response(status_code=codes.OK, url=url)
+
+    # Mock auth flow
+    def set_token(token: str) -> callable:
+        def inner(self):
+            self._token = token
+            self._expires = 0
+            yield from ()
+
+        return inner
+
+    auth = Auth(use_token_cache=False)
+    # Get token for the first time
+    auth.get_new_token_generator = MethodType(set_token(test_token), auth)
+    execute_generator_requests(auth.auth_flow(Request("GET", url)))
+    assert auth.token == test_token, "invalid access token"
+    assert auth.expired
+
+    # Refresh token
+    auth.get_new_token_generator = MethodType(set_token(test_token2), auth)
+    execute_generator_requests(auth.auth_flow(Request("GET", url)))
+    assert auth.token == test_token2, "Expired access token was not updated."
+
+
+def test_auth_uses_same_token_if_valid(
+    httpx_mock: HTTPXMock, test_token: str, test_token2: str
+) -> None:
+    """Auth reuses the token until it's expired."""
+    url = "https://host"
+    httpx_mock.add_response(status_code=codes.OK, url=url)
+
+    # Mock auth flow
+    def set_token(token: str) -> callable:
+        def inner(self):
+            self._token = token
+            self._expires = 2**32
+            yield from ()
+
+        return inner
+
+    auth = Auth(use_token_cache=False)
+    # Get token for the first time
+    auth.get_new_token_generator = MethodType(set_token(test_token), auth)
+    execute_generator_requests(auth.auth_flow(Request("GET", url)))
+    assert auth.token == test_token, "invalid access token"
+    assert not auth.expired
+
+    # Refresh token
+    auth.get_new_token_generator = MethodType(set_token(test_token2), auth)
+    execute_generator_requests(auth.auth_flow(Request("GET", url)))
+    assert auth.token == test_token, "Should not update token until it expires."
+
+
+def test_auth_adds_header(test_token: str) -> None:
+    """Auth adds required authentication headers to httpx.Request."""
+    auth = Auth(use_token_cache=False)
+    auth._token = test_token
+    auth._expires = 2**32
+    flow = auth.auth_flow(Request("get", ""))
+    request = next(flow)
+
+    assert "authorization" in request.headers, "missing authorization header"
+    assert (
+        request.headers["authorization"] == f"Bearer {test_token}"
+    ), "missing authorization header"
+
+
+@mark.nofakefs
+def test_auth_token_storage(
+    httpx_mock: HTTPXMock,
+    test_username: str,
+    test_password: str,
+    test_token,
+) -> None:
+    # Hack to replace Auth._token_cache attribute
+    class MAuth(Auth):
+        @property
+        def _token_storage(self):
+            return TokenSecureStorage(test_username, test_password)
```

Comment:
Would `unittest.mock.patch` be a better way of doing this?

Reply:
Yeah, didn't think of it for some reason